src/object/srv_obj.c

```diff
@@ -2387,6 +2414,39 @@ ds_obj_rw_handler(crt_rpc_t *rpc)
 		D_GOTO(out, rc);
 	}
 
+	/* XXX: For non-solo DTX, the leader and non-leader will make each own
+	 *	local modification in parallel. If the non-leader goes so fast
+	 *	as to when non-leader has already moved to handle next requet,
+	 *	the leader may has not really started current modification yet,
+	 *	such as being blocked at bulk transfer phase. Under such case,
+	 *	it is possible that when the non-leader handles next request,
+	 *	it hits the DTX that is just prepared locally, then non-leader
+	 *	will check such DTX status with leader. But at that time, the
+	 *	DTX entry on the leader does not exist, that will misguide the
+	 *	non-leader as missed to abort such DTX. To avoid such bad case,
+	 *	the leader need to build its DTX entry in DRAM before dispatch
+	 *	current request to non-leader.
+	 *
+	 *	On the other hand, even if for solo DTX, the PRC may be delay
+	 *	processed because of server side heavy load. If it is a update
+	 *	RPC with large data transfer, then it is possible that client
+	 *	think the update RPC is timeout and resend the RPC during the
+	 *	original RPC bulk data transfer that will yield CPU, and then
+	 *	the resend logic on server will not find related DTX entry since
+	 *	the DTX for original RPC is not 'prepared' yet. Under such case,
+	 *	the update request will be double executed on the server. That
+	 *	should be avoided. So pre-allocating DTX entry before bulk data
+	 *	transfer is necessary.
+	 */
+
+	if (!(flags & ORF_RESEND) &&
```

Comment:
Hmm, if the leader never get the request, it still needs to create the DRAM structure here?  Probably you need also lookup the DRAM structure table.

Reply:
If the RPC has set "RESED" but the IO logic does not find related DTX entry, then the parameter "flag" will NOT be set as "RESEND". So "if (!(flags & ORF_RESEND)" means that the real situation is that there is no related DTX entry. So if the original RPC is never received by the server, the subsequent resend RPC will also pin the DTX entry.

Reply:
Ah, I see. thanks.